platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/DefaultConfigurationCacheIO.kt

```diff
@@ -199,94 +206,141 @@ class DefaultConfigurationCacheIO internal constructor(
-                    action(stringDecoder)
-                }
+    fun sharedObjectDecoder(baseFile: ConfigurationCacheStateFile, globalsFile: ConfigurationCacheStateFile, stringDecoder: StringDecoder): SharedObjectDecoder =
+        isUsingShareableObjectSharingStrategy(baseFile).let { deduplicate ->
+            if (deduplicate) {
+                // Create a context that honors global value duplication
+                // but uses an inline global value decoder
+                val (globalContext, _) = readContextFor(globalsFile, stringDecoder)
+                globalContext.push(IsolateOwners.OwnerGradle(host.currentBuild.gradle))
+                DefaultSharedObjectDecoder { globalContext }
+            } else {
+                InlineSharedObjectDecoder
             }
         }
 
+    private
+    fun <T> withStringEncoderFor(stateFile: ConfigurationCacheStateFile, action: (StringEncoder) -> T): T =
+        stringsFileFor(stateFile).let { stringsFile ->
+            stringEncoderFor(stringsFile).use(action)
+        }
+
+    private
+    fun <T> withStringDecoderFor(stateFile: ConfigurationCacheStateFile, action: (StringDecoder) -> T): T =
+        stringsFileFor(stateFile).let { stringsFile ->
+            stringDecoderFor(stringsFile).use(action)
+        }
+
+    private
+    fun <T> withSharedObjectEncoderFor(baseFile: ConfigurationCacheStateFile, stringEncoder: StringEncoder, action: (SharedObjectEncoder) -> T): T =
+        sharedObjectFileFor(baseFile).let { globalsFile ->
+            sharedObjectEncoderFor(baseFile, globalsFile, stringEncoder).use(action)
+        }
+
+    private
+    fun <T> withSharedObjectDecoderFor(baseFile: ConfigurationCacheStateFile, stringDecoder: StringDecoder, action: (SharedObjectDecoder) -> T): T =
+        sharedObjectFileFor(baseFile).let { globalsFile ->
+            sharedObjectDecoder(baseFile, globalsFile, stringDecoder).use(action)
+        }
+
     private
     fun stringsFileFor(stateFile: ConfigurationCacheStateFile) =
         stateFile.relatedStateFile(Path.path(".strings"))
 
     private
-    fun <T> readConfigurationCacheStateWithStringDecoder(
+    fun sharedObjectFileFor(stateFile: ConfigurationCacheStateFile) =
+        stateFile.stateFileForSharedObjects()
+
+    private
+    fun <T> readConfigurationCacheStateWithSpecialDecoders(
         stringEncoder: StringDecoder,
+        sharedObjectDecoder: SharedObjectDecoder,
         stateFile: ConfigurationCacheStateFile,
         action: suspend MutableReadContext.(ConfigurationCacheState) -> T
-    ) = withReadContextFor(stateFile, stringEncoder) { codecs ->
+    ) = withReadContextFor(stateFile, stringEncoder, sharedObjectDecoder) { codecs ->
         ConfigurationCacheState(codecs, stateFile, ChildContextSource(stateFile), eventEmitter, host).run {
             action(this)
         }
     }
 
     private
-    fun <T> writeConfigurationCacheStateWithStringEncoder(
+    fun <T> writeConfigurationCacheStateWithCustomEncoders(
```

Comment:
💄 `Custom` or `Special`?

Reply:
I had `Special` at some point, will update

Reply:
of course, that is why it is inconsistent...